src/swap.ts

```diff
@@ -0,0 +1,284 @@
+  ? number extends K1
+    ? ReadonlyArray<T[number]>
+    : SafeIndex<T, K1> extends never
+    ? T
+    : K2 extends number
+    ? number extends K2
+      ? ReadonlyArray<T[number]>
+      : SafeIndex<T, K2> extends never
+      ? T
+      : SwapArrayInternal<T, SafeIndex<T, K1>, SafeIndex<T, K2>>
+    : T
+  : T;
+
+type Swap<
+  T extends ReadonlyArray<unknown> | string,
+  K1 extends number,
+  K2 extends number
+> = T extends string
+  ? string extends T
+    ? string
+    : SwapString<T, K1, K2>
+  : T extends IterableContainer
+  ? SwapArray<T, K1, K2>
+  : never;
+
+/**
+ * Swaps the positions of two elements in a list, string or an object based on the provided keys.
+ * For strings, the keys represent the index positions of the characters to be swapped.
+ * For objects, the keys represent the property keys of the properties to be swapped.
+ *
+ * @param data the item to be manipulated. This can be an array, or a string.
+ * @param key1 the first index
+ * @param key2 the second index
+ *
+ * @signature
+ *   swap(key1, key2)(data)
+ *
+ * @example
+ *   swap(['a', 'b', 'c'], 0, 1) // => ['b', 'a', 'c']
+ *   swap('abc', 0, 1) // => 'bac'
+ *
+ * @category Array
+ *
+ * @returns Returns the manipulated array or string.
+ *
+ * @data_first
+ */
+export function swap<
+  T extends ReadonlyArray<unknown> | string,
+  K1 extends keyof T & number,
+  K2 extends keyof T & number
+>(data: T, key1: K1, key2: K2): Swap<T, K1, K2>;
+/**
+ * @param data the object to be manipulated
+ * @param key1 the first property key
+ * @param key2 the second property key
+ *
+ * @signature
+ *   swap(key1, key2)(data)
+ *
+ * @returns Returns the manipulated object.
+ *
+ * @example
+ *   swap({a: 1, b: 2, c: 3}, 'a', 'b') // => {a: 2, b: 1, c: 3}
+ */
+export function swap<T extends object, K1 extends keyof T, K2 extends keyof T>(
+  data: T,
+  key1: K1,
+  key2: K2
+): ObjectSwap<T, K1, K2>;
+
+/**
+ * @param key1 the first index
+ * @param key2 the second index
+ *
+ * @signature
+ *   swap(key1, key2)(data)
+ *
+ * @example
+ *   swap(0, 1)(['a', 'b', 'c']) // => ['b', 'a', 'c']
+ *   swap(0, -1)('abc') // => 'cba'
+ *
+ * @category Array
+ * @returns Returns the manipulated array or string.
+ * @data_last
+ */
+export function swap<K1 extends number, K2 extends number>(
+  key1: K1,
+  key2: K2
+): <T extends ReadonlyArray<unknown> | string>(data: T) => Swap<T, K1, K2>;
+
+/**
+ * Swaps the positions of two properties in an object based on the provided keys.
+ * @param key1 the first property key
+ * @param key2 the second property key
+ *
+ * @signature
+ *   swap(key1, key2)(data)
+ *
+ * @example
+ *   swap('a', 'b')({a: 1, b: 2, c: 3}) // => {a: 2, b: 1, c: 3}
+ *
+ * @returns Returns the manipulated object.
+ */
+export function swap<K1 extends PropertyKey, K2 extends PropertyKey>(
+  key1: K1,
+  key2: K2
+): <T extends { [P in K1 | K2]: any }>(data: T) => ObjectSwap<T, K1, K2>;
```

Comment:
This is an old remanent, it should be modernized like the data-first impl
```suggestion
export function swap<T extends object, K1 extends keyof T, K2 extends keyof T>(
  key1: K1,
  key2: K2
): (data: T) => ObjectSwap<T, K1, K2>;
```